Add docs on custom Operations (#3507)
airbytehq/airbyte: docs/understanding-airbyte/connections/README.md

# Connections

A connection is a configuration for syncing data between a source and a destination. To setup a connection, a user must configure things such as:

* A catalog selection: which [streams and fields](../catalog.md) to replicate from the source
* Sync mode: how streams should be replicated (read and write):
* Sync schedule: when to trigger a sync of the data
* Optional transformations: how to convert Airbyte protocol messages (raw JSON blob) data into some other data representations. 

## Sync modes

> A sync mode governs how Airbyte reads from a source and writes to a destination. Airbyte provides different sync modes to account for various use cases. To minimise confusion, a mode's behavior is reflected in its name. The easiest way to understand Airbyte's sync modes is to understand how the modes are named.

Reviewer: More anglicization :)
```suggestion
A sync mode governs how Airbyte reads from a source and writes to a destination. Airbyte provides different sync modes to account for various use cases. To minimize confusion, a mode's behavior is reflected in its name. The easiest way to understand Airbyte's sync modes is to understand how the modes are named.
```

  * Full Refresh: Read everything in the source.

2. The second part of the sync mode name denotes how the destination connector writes data. This is not affected by how the source connector produced the data:

  * Overwrite: Overwrite by first deleting existing data in the destination.
  * Append: Write by adding data to existing tables in the destination.
  * Deduped History: Write by first adding data to existing tables in the destination to keep a history of changes. The final table is produced by de-duplicating the intermediate ones using a primary key.

A sync mode is therefore, a combination of a source and destination mode together. The UI exposes the following options, whenever both source and destination connectors are capable to support it for the corresponding stream:
* [Full Refresh Overwrite](full-refresh-overwrite.md): Sync the whole stream and replace data in destination by overwriting it.
* [Full Refresh Append](full-refresh-append.md): Sync the whole stream and append data in destination.
* [Incremental Append](incremental-append.md): Sync new records from stream and append data in destination.

> * [Incremental Deduped History](incremental-deduped-history.md): Sync new records from stream and append data in destination, also provides a de-duplicated view mirroring the state of the stream in the source.

Reviewer 2: @ChristopheDuong can add a note saying that for the first run **Incremental** mode works as a full refresh ?

# Connections

A connection is a configuration for syncing data between a source and a destination. To setup a connection, a user must configure things such as:

* A catalog selection: which [streams and fields](../catalog.md) to replicate from the source
* Sync mode: how streams should be replicated (read and write):
* Sync schedule: when to trigger a sync of the data
* Optional transformations: how to convert Airbyte protocol messages (raw JSON blob) data into some other data representations. 

## Sync modes

Note that Sync mode names are generally composed of two parts on purpose. This is because Source and Destination connectors are decoupled from each other.
One is responsible for reading, and the other for writing.

1. The first part of the sync mode name denotes how source connectors work:
  

>   * Incremental: Read only the last records in the source since last read operation during a previous sync

Reviewer 3: ```suggestion
  * Incremental: Read records added to the source since the last sync job.
```

>       * Method 1: comparing to a cursor value

Reviewer 3: ```suggestion
      * Method 1: Using a cursor. Generally supported by all sources.
```

>       * Method 2: using [CDC](../cdc.md), only available for certain limited number of sources that can support it.

Reviewer 3: ```suggestion
      * Method 2: Method 2: Using change data capture. Only supported by some sources. See [CDC](../cdc.md) for more info.
```

>   * Full Refresh: Read everything in the source

Reviewer 3: ```suggestion
  * Full Refresh: Read everything in the source.
```

> 2. The second part of the sync mode name denotes how destination connectors work, regardless of how the source is producing data:

Reviewer 3: ```suggestion
2. The second part of the sync mode name denotes how the destination connector writes data. This is not affected by how the source connector produced the data:
```

>   * Overwrite: Delete and then write over data in the destination that already exists

Reviewer 3: ```suggestion
  * Overwrite: Delete and then overwrite existing data in the destination.
```

>   * Append: Write by adding data to existing tables in the destination

Reviewer 3: ```suggestion
  * Append: Write by adding data to existing tables in the destination.
```

>   * Deduped History: Write by adding data to existing tables in the destination to keep a history of changes. The final table is produced by deduplicating the intermediate ones using a primary key.

Reviewer 3: ```suggestion
  * Deduped History: Write by first adding data to existing tables in the destination to keep a history of changes. The final table is produced by de-duplicating the intermediate ones using a primary key.
```

> A sync mode is therefore, a combination of a source and destination mode together. The UI exposes the following options, whenever both source and destination connectors are capable to support it for the corresponding stream:

Reviewer 3: this is great!

Reviewer: +1

Reviewer 2: @avaidyanatha could we add some illustrations about the sync modes? This can help users understand the concept easier.
![image](https://user-images.githubusercontent.com/5154322/120338780-ee2c6000-c2ca-11eb-9497-1c732bb3f7f2.png)

* **October 3rd, 2:01pm:** since the last sync was less than 24 hours ago, no sync is run
* **October 3rd, 5:01pm:** It has been more than 24 hours since the last sync, so a sync is run

## Optional operations

### Airbyte basic normalization

As described by the [Airbyte Protocol from the Airbyte Specifications](../airbyte-specification.md), a replication is composed of source connectors that are transmitting data in a JSON format. It is then written as such by the destination connectors. 

On top of this replication, Airbyte provides the option to enable or disable an additional transformation step at the end of the sync called [basic normalization](../basic-normalization.md). This operation is:

> - only available for destinations that supports dbt execution.

Reviewer: ```suggestion
- only available for destinations that support dbt execution.
```

- responsible for automatically generating a pipeline or a DAG of dbt transformation models to convert JSON blob objects into normalized tables.

> - running and applying these dbt models to the data written in the destination.

Reviewer: ```suggestion
- responsible for running and applying these dbt models to the data written in the destination.
```

* **October 3rd, 2:01pm:** since the last sync was less than 24 hours ago, no sync is run
* **October 3rd, 5:01pm:** It has been more than 24 hours since the last sync, so a sync is run

## Optional operations

### Airbyte basic normalization

As described by the [Airbyte Protocol from the Airbyte Specifications](../airbyte-specification.md), a replication is composed of source connectors that are transmitting data in a JSON format. It is then written as such by the destination connectors. 

On top of this replication, Airbyte provides the option to enable or disable an additional transformation step at the end of the sync called [basic normalization](../basic-normalization.md). This operation is:

- only available for destinations that supports dbt execution.
- responsible for automatically generating a pipeline or a DAG of dbt transformation models to convert JSON blob objects into normalized tables.
- running and applying these dbt models to the data written in the destination.

### Custom sync operations

> Further operations can be included in a sync on top of Airbyte basic normalization (or even to replace it completely). Potential applications are:

Reviewer 3: I think it's better to point to the operations.md file, and move the list describing the operations there. What do you think?
```suggestion
Further operations can be included in a sync on top of Airbyte basic normalization (or even to replace it completely). See operations.md.
```

> - Customized normalization to better fit the requirements of your own business context.

Reviewer 3: So move this to the operations.md intro.
